lib/puppet/type/postgresql_psql.rb

```diff
@@ -135,11 +135,7 @@ def matches(value)
   end
 
   autorequire(:anchor) do
-    ["postgresql::server::service::begin::#{self[:instance]}"]
-  end
-
-  autorequire(:service) do
```

Comment:
I'm trying to understand the logic. Don't you **always** need the server to be running to execute any SQL statements?

Reply:
Oh yes, that makes sense. Though with the newest versions (Fedora 38 with PostgreSQL 15) I see the systemd unit is `Type=notify` so I'd assume that there it will only really be ready once it's listening (spoiler: the patch sends `READY=1` right after logging `database system is ready to accept connections`).

Digging into this, it was introduced in https://github.com/postgres/postgres/commit/7d17e683fcc28a1b371c7dd02935728cd2cbf9bf so with PostgreSQL 9.6 it became possible to support. Looking at EL8 that's built with `Type=notify`, but EL7 probably isn't. Sadly Debian 11 & 12 are also `Type=forking`. That means you're right and we still need to depend on the connection validator.

Reply:
> Oh yes, that makes sense. Though with the newest versions (Fedora 38 with PostgreSQL 15) I see the systemd unit is `Type=notify` so I'd assume that there it will only really be ready once it's listening (spoiler: the patch sends `READY=1` right after logging `database system is ready to accept connections`).
> 
> Digging into this, it was introduced in [postgres/postgres@7d17e68](https://github.com/postgres/postgres/commit/7d17e683fcc28a1b371c7dd02935728cd2cbf9bf) so with PostgreSQL 9.6 it became possible to support. Looking at EL8 that's built with `Type=notify`, but EL7 probably isn't. Sadly Debian 11 & 12 are also `Type=forking`. That means you're right and we still need to depend on the connection validator.

These seem to be the relevant bugs about this:

* https://bugs.launchpad.net/ubuntu/+source/postgresql-common/+bug/1862138
* https://bugs.debian.org/872915